ghcide/src/Development/IDE/Plugin/TypeLenses.hs

```diff
@@ -340,3 +372,165 @@ pprPatSynTypeWithoutForalls p = pprPatSynType pWithoutTypeVariables
     builder = patSynBuilder p
     field_labels = patSynFieldLabels p
     orig_args' = map scaledThing orig_args
+
+-- --------------------------------------------------------------------------------
+
+-- | A binding expression with its id and location.
+data WhereBinding = WhereBinding
+    { bindingId  :: Id
+    -- ^ Each WhereBinding represents an id in binding expression.
+    , bindingLoc :: SrcSpan
+    -- ^ Location for an individual binding in a pattern.
+    -- Here we use the 'bindingLoc' and offset to render the type signature at the proper place.
+    , offset     :: Int
+    -- ^ Column offset between whole binding and individual binding in a pattern.
+    --
+    -- Example: For @(a, b) = (1, True)@, there will be two `WhereBinding`s:
+    -- - `a`: WhereBinding id_a loc_a 0
+    -- - `b`: WhereBinding id_b loc_b 4
+    }
+
+-- | Existing bindings in a where clause.
+data WhereBindings = WhereBindings
+  { bindings         :: [WhereBinding]
+  , existingSigNames :: [Name]
+  -- ^ Names of existing signatures.
+  -- It is used to hide type lens for existing signatures.
+  --
+  -- NOTE: The location of this name is equal to
+  -- the binding name.
+  --
+  -- Example:
+  -- @
+  -- f :: Int
+  -- f = 42
+  -- @
+  -- The location of signature name `f`(first line) is equal to
+  -- the definition of `f`(second line).
+  }
+
+data GetWhereBindingTypeSigs = GetWhereBindingTypeSigs
+  deriving (Generic, Show, Eq, Ord, Hashable, NFData)
+
+type BindingSigMap = Map.Map Id String
+
+newtype WhereBindingTypeSigsResult = WhereBindingTypeSigsResult ([WhereBindings], BindingSigMap)
+
+instance Show WhereBindingTypeSigsResult where
+  show _ = "<GetTypeResult.where>"
+
+instance NFData WhereBindingTypeSigsResult where
+  rnf = rwhnf
+
+type instance RuleResult GetWhereBindingTypeSigs = WhereBindingTypeSigsResult
+
+whereBindingRules :: Recorder (WithPriority Log) -> Rules ()
+whereBindingRules recorder = do
+  define (cmapWithPrio LogShake recorder) $ \GetWhereBindingTypeSigs nfp -> do
+    tmr <- use TypeCheck nfp
+    -- we need session here for tidying types
+    hsc <- use GhcSession nfp
+    result <- liftIO $ whereBindingType (tmrTypechecked <$> tmr) (hscEnv <$> hsc)
+    pure ([], result)
+
+whereBindingType :: Maybe TcGblEnv -> Maybe HscEnv -> IO (Maybe WhereBindingTypeSigsResult)
+whereBindingType (Just gblEnv) (Just hsc) = do
+  let wheres = findWhereQ (tcg_binds gblEnv)
```

Comment:
I wonder if we should just go ahead and do lets at the same time? I would imagine it's very similar code?